Sources/SwiftSyntaxBuilder/BuildableNodes.swift.gyb

```diff
@@ -106,7 +107,32 @@ public struct ${type.buildable()}: ${base_type.buildable()}, ${type.expressible_
 
   /// Conformance to `${base_type.buildable()}`.
   public func build${base_type.base_name()}(format: Format, leadingTrivia: Trivia? = nil) -> ${base_type.syntax()} {
+    %   if type.buildable() != 'SourceFile':
     let result = build${type.base_name()}(format: format, leadingTrivia: leadingTrivia)
+    % end
+    %   if type.buildable() == 'SourceFile':
+    let indentedLines = itemLeadingTriviaPieces
+      .map { Trivia(pieces: [$0, .newlines(1)]) }
+      .reduce(.zero, +)
+    
+    let combinedTrivia = [format._makeIndent(),
+                          leadingTrivia,
+                          indentedLines]
+      .compactMap { $0 }
+      .reduce(.zero, +)
+    %   if type.buildable() == 'SourceFile':
+    let result = buildSourceFile(format: format, leadingTrivia: combinedTrivia)
+    % end
+    % else:
+              
+      if !itemLeadingTriviaPieces.isEmpty {
+        let indentedLeadingTrivia = itemLeadingTriviaPieces
+          .compactMap { Trivia(pieces:[ $0, .newlines(1)]) }
+          .map { $0 + format._makeIndent() }
+          .reduce(.zero, +)
+        return ${base_type.syntax()}(result).withLeadingTrivia(indentedLeadingTrivia)
+      }
+    %   end
```

Comment:
I don’t think the special handling for `SourceFile` is sufficient here. The same issue will arise for any syntax node that doesn’t have a token before its first child node (I’m happy to be proven wrong by a test case). For example we should see the same issue if you try to create a `TypealiasDecl` like the following where both the `private` attribute and the `TypealiasDecl` itself have comments attached to it.

```swift
/// This comment is added to the TypealiasDecl
/// This comment is added to the `private` keyword
private typealias MyInt = Int
```

Is there any issue with applying the solution you have implemented for `SourceFile` to all syntax nodes (I haven’t checked).

---

Also, I think there’s a few `end` missing and the ones that do exist have the wrong indentation level sometimes. Could you make sure that every if-statement has a matching `end` on the same indentation level?